// @crtxio/hashers/dist/sha3.mjs
import { GenericHasher } from './generic.mjs';
import * as sha3 from '@noble/hashes/sha3';
import './abstract.mjs';

const sha3224 = GenericHasher.from({
  code: 23,
  name: "sha3-224",
  encoder: sha3.sha3_224
});
const sha3256 = GenericHasher.from({
  code: 22,
  name: "sha3-256",
  encoder: sha3.sha3_256
});
const sha3384 = GenericHasher.from({
  code: 21,
  name: "sha3-384",
  encoder: sha3.sha3_384
});
const sha3512 = GenericHasher.from({
  code: 20,
  name: "sha3-512",
  encoder: sha3.sha3_512
});
const keccak224 = GenericHasher.from({
  code: 26,
  name: "keccak-224",
  encoder: sha3.keccak_224
});
const keccak256 = GenericHasher.from({
  code: 27,
  name: "keccak-256",
  encoder: sha3.keccak_256
});
const keccak384 = GenericHasher.from({
  code: 28,
  name: "keccak-384",
  encoder: sha3.keccak_384
});
const keccak512 = GenericHasher.from({
  code: 29,
  name: "keccak-512",
  encoder: sha3.keccak_512
});
const shake128 = GenericHasher.from({
  code: 24,
  name: "shake-128",
  encoder: (data) => {
    return sha3.shake128(data, { dkLen: 32 });
  }
});
const shake256 = GenericHasher.from({
  code: 25,
  name: "shake-256",
  encoder: (data) => {
    return sha3.shake256(data, { dkLen: 64 });
  }
});

export { keccak224, keccak256, keccak384, keccak512, sha3224, sha3256, sha3384, sha3512, shake128, shake256 };
