// @crtxio/bases/dist/base32.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const rfc4648 = require('./rfc4648.cjs');
require('./abstract.cjs');

const base32 = rfc4648.RFC4648.from({
  prefix: "b",
  name: "base32",
  alphabet: "abcdefghijklmnopqrstuvwxyz234567",
  bits: 5
});
const base32upper = rfc4648.RFC4648.from({
  prefix: "B",
  name: "base32upper",
  alphabet: "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567",
  bits: 5
});
const base32pad = rfc4648.RFC4648.from({
  prefix: "c",
  name: "base32pad",
  alphabet: "abcdefghijklmnopqrstuvwxyz234567=",
  bits: 5
});
const base32padupper = rfc4648.RFC4648.from({
  prefix: "C",
  name: "base32padupper",
  alphabet: "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567=",
  bits: 5
});
const base32hex = rfc4648.RFC4648.from({
  prefix: "v",
  name: "base32hex",
  alphabet: "0123456789abcdefghijklmnopqrstuv",
  bits: 5
});
const base32hexupper = rfc4648.RFC4648.from({
  prefix: "V",
  name: "base32hexupper",
  alphabet: "0123456789ABCDEFGHIJKLMNOPQRSTUV",
  bits: 5
});
const base32hexpad = rfc4648.RFC4648.from({
  prefix: "t",
  name: "base32hexpad",
  alphabet: "0123456789abcdefghijklmnopqrstuv=",
  bits: 5
});
const base32hexpadupper = rfc4648.RFC4648.from({
  prefix: "T",
  name: "base32hexpadupper",
  alphabet: "0123456789ABCDEFGHIJKLMNOPQRSTUV=",
  bits: 5
});
const base32z = rfc4648.RFC4648.from({
  prefix: "h",
  name: "base32z",
  alphabet: "ybndrfg8ejkmcpqxot1uwisza345h769",
  bits: 5
});

exports.base32 = base32;
exports.base32hex = base32hex;
exports.base32hexpad = base32hexpad;
exports.base32hexpadupper = base32hexpadupper;
exports.base32hexupper = base32hexupper;
exports.base32pad = base32pad;
exports.base32padupper = base32padupper;
exports.base32upper = base32upper;
exports.base32z = base32z;
