// @crtxio/bases/dist/base32.mjs
import { RFC4648 } from './rfc4648.mjs';
import './abstract.mjs';

const base32 = RFC4648.from({
  prefix: "b",
  name: "base32",
  alphabet: "abcdefghijklmnopqrstuvwxyz234567",
  bits: 5
});
const base32upper = RFC4648.from({
  prefix: "B",
  name: "base32upper",
  alphabet: "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567",
  bits: 5
});
const base32pad = RFC4648.from({
  prefix: "c",
  name: "base32pad",
  alphabet: "abcdefghijklmnopqrstuvwxyz234567=",
  bits: 5
});
const base32padupper = RFC4648.from({
  prefix: "C",
  name: "base32padupper",
  alphabet: "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567=",
  bits: 5
});
const base32hex = RFC4648.from({
  prefix: "v",
  name: "base32hex",
  alphabet: "0123456789abcdefghijklmnopqrstuv",
  bits: 5
});
const base32hexupper = RFC4648.from({
  prefix: "V",
  name: "base32hexupper",
  alphabet: "0123456789ABCDEFGHIJKLMNOPQRSTUV",
  bits: 5
});
const base32hexpad = RFC4648.from({
  prefix: "t",
  name: "base32hexpad",
  alphabet: "0123456789abcdefghijklmnopqrstuv=",
  bits: 5
});
const base32hexpadupper = RFC4648.from({
  prefix: "T",
  name: "base32hexpadupper",
  alphabet: "0123456789ABCDEFGHIJKLMNOPQRSTUV=",
  bits: 5
});
const base32z = RFC4648.from({
  prefix: "h",
  name: "base32z",
  alphabet: "ybndrfg8ejkmcpqxot1uwisza345h769",
  bits: 5
});

export { base32, base32hex, base32hexpad, base32hexpadupper, base32hexupper, base32pad, base32padupper, base32upper, base32z };
